### PR TITLE
otInstance Declarations

### PR DESCRIPTION
**Pull request by nibanks**
This separates out the PR for refactoring globals/statics into a single structure, to just include API changes.

### CODE REVIEW

**Comment by jwhui** [comment]
Thanks for this PR.  At first glance, I think break the API changes out separately is the right incremental approach.  I will review this in detail soon.

File: include/openthread.h [hunk 1]
```diff
@@ -154,87 +160,140 @@ extern void otSignalTaskletPending(void);
...
+ * called before any other calls to OpenThread. By default, OpenThread is initialized in the 'enabled' state.
+ *
+ * @param[in]    aInstanceBuffer      The buffer for OpenThread to use for allocating the otInstance structure.
+ * @param[inout] aInstanceBufferSize  On input, the size of aInstanceBuffer. On output, if not enough space for otInstance,
+                                      the number of bytes required for otInstance.
+ *
+ * @retval otInstance*  The new OpenThread instance structure.
+ *
+ */
+otInstance *otInstanceInit(void *aInstanceBuffer, uint64_t *aInstanceBufferSize);
```
Comment (jwhui) [inline include/openthread.h]: The `otEnable()` and `otDisable()` APIs were really intended to serve this purpose.  In particular, `otEnable()` MUST be called before any other calls to OpenThread.  `otDisable()` may be called when the client no longer wishes to use OpenThread.  After calling `otDisable()` a call to `otEnable()` must be made before any further calls to OpenThread are mode.

Any concerns if we simply make `otInstanceInit()` and `otInstanceFinalize()` to be the new `otEnable()` and `otDisable()`?

File: include/openthread-types.h [hunk 2]
```diff
@@ -37,13 +37,19 @@
...
 #include <platform/toolchain.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 /**
+ * This type represents the OpenThread instance structure.
+ */
+typedef struct otInstance otInstance;
```
Comment (jwhui) [inline include/openthread-types.h]: We should avoid having clients of OpenThread needing to pull in definitions for internal structures.  Can we just typedef a `void` instead?

**Comment by jwhui** [comment]
This looks good overall.  See my comments above.

File: include/openthread.h (hunk 1 above)
Comment (nibanks) [inline include/openthread.h]: I'm fine with that. I assume we would just get rid of the mEnabled state then, right?

Comment (jwhui) [inline include/openthread.h]: Yes, I agree with removing `mEnabled`.

File: include/openthread-types.h (hunk 2 above)
Comment (nibanks) [inline include/openthread-types.h]: This doesn't require the caller to have any actual knowledge of the implementation. All my Windows driver code uses it just fine and has no internal file references.

Comment (jwhui) [inline include/openthread-types.h]: Okay, this WFM as long as this remains an incomplete type for the caller.

File: include/openthread.h (hunk 1 above)
Comment (darconeous) [inline include/openthread.h]: As an aside, I think we conflate initialization with enable/disable behavior all over the place in the existing API. "Initializing" something is conceptually different from "Enabling" something, especially when that something can then be "Disabled". But that is an issue that can be considered separately from this pull request.

File: include/openthread-types.h (hunk 2 above)
Comment (darconeous) [inline include/openthread-types.h]: I've done this trick a few times myself. It's rather convenient since it allows you to avoid casting.

**Comment by darconeous** [comment]
Reviewed. LGTM. 👍

**Comment by jwhui** [comment]
LGTM 👍

**Pull request merged**